pkg/logbridge/logbridge_test.go

```diff
@@ -0,0 +1,161 @@
+package logbridge
+
+import (
+	"bytes"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/square/p2/pkg/logging"
+)
+
+const newLine = byte(10)
+
+type TrackingWriter struct {
+	bytes     []byte
+	numWrites int
+}
+
+func (tw *TrackingWriter) Write(p []byte) (int, error) {
+	tw.numWrites++
+	return len(p), nil
+}
+
+func TestLogBridge(t *testing.T) {
+	type testCase struct {
+		inputSize      int
+		bridgeCapacity int
+		expected       int
+	}
+
+	testCases := []testCase{
+		{inputSize: 0, bridgeCapacity: 0, expected: 0},
+		{inputSize: 10, bridgeCapacity: 10, expected: 10},
+		{inputSize: 100, bridgeCapacity: 10, expected: 10},
+		{inputSize: 10, bridgeCapacity: 100, expected: 10},
+	}
+
+	for i, testCase := range testCases {
+		t.Logf("test case %d", i)
+		input := make([]byte, 0, 2*testCase.inputSize)
+		for i := 0; i < testCase.inputSize; i++ {
+			input = append(input, byte('a'), newLine)
+		}
+
+		reader := bytes.NewReader(input)
+		writer := &TrackingWriter{}
+
+		Copy(writer, reader, testCase.bridgeCapacity, logging.DefaultLogger)
+		if writer.numWrites < testCase.expected {
+			t.Errorf("Writer did not receive enough writes, got %d expected: %d", writer.numWrites, testCase.inputSize)
+		}
+	}
+}
+
+// This writer becomes latent after receiving capacity writes.
+type LatentWriter struct {
+	capacity int
+	writes   int
+	Bytes    []byte
+}
+
+func (sw *LatentWriter) Write(p []byte) (n int, err error) {
+	if sw.writes > sw.capacity {
+		time.Sleep(10 * time.Millisecond)
+	}
+	sw.writes++
+
+	sw.Bytes = append(sw.Bytes, p...)
```

Comment:
Like I mentioned above, might be worth implementing this as a wrapper around another writer. Then you only need to do the sleep here, and the tracking is provided by trackingwriter, and the bytes comes from bytes.Buffer.